main/src/library/Fixpoint/Stratifier.flix

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright 2021 Benjamin Dahse
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+use Fixpoint/Ast.{Datalog, PrecedenceGraph, PredSym};
+use Fixpoint/Ast.Datalog.Datalog;
+use Fixpoint/Ast.PrecedenceGraph.PrecedenceGraph;
+use Fixpoint/Ast.PrecedenceEdge.{PositiveEdge, NegativeEdge};
+
+namespace Fixpoint {
+    def stratify(d: Datalog[v]): Map[PredSym, Int32] = match d {
+        case Datalog(_, rules) =>
+            predicatesOf(rules) |>
+            Set.foldRight(p -> Map.insert(p, 0), Map#{}) |>
+            stratifyHelper(Fixpoint/Ast.mkDepGraph(d))
+    }
+
+    /// Compute a stratification (if one exists) with Ullman's algorithm.
```

Comment:
If there are only facts the precedence graph will be empty and hence the stratification is empty. That's not a problem because facts are compiled independently from the stratification, for other reasons.